crates/fiber/src/arch/aarch64.S

```diff
@@ -0,0 +1,111 @@
+// A WORD OF CAUTION
+//
+// This entire file basically needs to be kept in sync with itself. It's not
+// really possible to modify just one bit of this file without understanding
+// all the other bits. Documentation tries to reference various bits here and
+// there but try to make sure to read over everything before tweaking things!
+//
+// Also at this time this file is heavily based off the x86_64 file, so you'll
+// probably want to read that one as well.
+
+#define GLOBL(fnname) .globl fnname
+#define TYPE(fnname) .type fnname,@function
+#define FUNCTION(fnname) fnname
+#define SIZE(fnname) .size fnname,.-fnname
+
+// fn(top_of_stack(%x0): *mut u8)
+GLOBL(wasmtime_fiber_switch)
+.p2align 2
+TYPE(wasmtime_fiber_switch)
+FUNCTION(wasmtime_fiber_switch):
+    // Save all callee-saved registers on the stack since we're assuming
+    // they're clobbered as a result of the stack switch.
+    str lr, [sp, -16]!
+    stp x20, x19, [sp, -16]!
+    stp x22, x21, [sp, -16]!
+    stp x24, x23, [sp, -16]!
+    stp x26, x25, [sp, -16]!
+    stp x28, x27, [sp, -16]!
```

Comment:
The bottom 64 bits of `v8` -- `v15` (aka `q8` -- `q15`) are callee-saved as well (yeah, it's kinda weird!). `stp q8, q9, ...`, etc. should work.

Reply:
I think this is the first time I've done aarch64 assembly, so want to confirm something to make sure. According [to this](https://developer.arm.com/architectures/learn-the-architecture/aarch64-instruction-set-architecture/registers-in-aarch64-general-purpose-registers) the low 64-bits of the vector registers are `dN`, does that sound right? From what [this is saying though](https://developer.arm.com/architectures/learn-the-architecture/aarch64-instruction-set-architecture/procedure-call-standard) is it true that `d8` to `d31` all need to be saved? (wow that's a lot of registers)

Reply:
Err, sorry, `d8` -- `d15`, yes. That document is a bit perplexing to me, though; the AArch64 "AAPCS" (standard ABI) ([direct pdf link](https://c9x.me/compile/bib/abi-arm64.pdf) to ABI doc, or [ARM's download page](https://developer.arm.com/documentation/ihi0055/b/)) says in 5.1.2 that `v0` -- `v7` and `v16` -- `v31` are caller-saved, and that's what we've implemented elsewhere (and what our friendly local ARM engineers have pointed us to -- see e.g. #2254).

I'm actually wondering if the ARM page you've linked has a typo; they segment the registers into 8-15 and 16-31, but write "callee-save" for both; there'd be no reason to write the two ranges separately otherwise. In any case, hopefully the above helps somewhat!

Reply:
Oh interesting! That does indeed look like a typo. Ok cool, then I think this should be handled in e40bd97